### PR TITLE
Add unit tests for Utils class #107

### LINKED ISSUE #107: Add unit tests for Utils class
as per title

### PR DESCRIPTION
**Pull request by zzzzwen**
Fixes #107
Ready for review

### CODE REVIEW

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            return new Tag(tagName);
+        } catch (IllegalValueException e) {
+            throw new RuntimeException("test tag name should be valid", e);
+        }
+    }
+
+    @Test
+    public void isAnyNull_emptyInput_returnsFalse() {
+        assertFalse(Utils.isAnyNull());
+    }
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: `isAnyNull` is a simple method. Can be tested using one test method. Cases to test:
No nulls: 
* any non-empty list, 
* an empty list
With nulls: 
* non empty list with just one at the beginning (to check for bugs missing the first item in the list), 
* list with multiple objects, couple of nulls in the middle
* non empty list with one null as the last element (to check for bugs missing last item in the list)

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        assertFalse(Utils.isAnyNull(emptyList, "A string"));
+    }
+
+    @Test
+    public void elementsAreUnique_emptyCollection_returnsTrue() {
+        assertTrue(Utils.elementsAreUnique(new ArrayList<>()));
+    }
+
+    @Test
+    public void elementsAreUnique_singleElement_returnsTrue() {
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Can it be consolidated like this?
```java
@Test
public void elementsAreUnique() {
   //create test objects obj1, obj2, obj3, obj1copy
    assertAreUnique() //empty list
    assertAreUnique(obj1); //only one object
    assertAreUnique(obj1, obj2, obj3); // all objects unique
    assertNotUnique(obj1, obj2, obj1copy); //...
}

void assertAreUnique(Object... objects){
    ...
}
```

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 1 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: noted

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 2 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: noted

**Comment by zzzzwen** [comment]
Updated according to reviews

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        tag1 = buildTag("tag1");
+        tag1Copy = buildTag("tag1");
+        tag2 = buildTag("tag2");
+    }
+
+    private static Tag buildTag(String tagName) {
+        try {
+            return new Tag(tagName);
+        } catch (IllegalValueException e) {
+            throw new RuntimeException("test tag name should be valid", e);
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: You can use `fail()` here instead.

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    private static Tag tag1;
+    private static Tag tag1Copy;
+    private static Tag tag2;
+
+    @Before
+    public void setup() {
+        tag1 = buildTag("tag1");
+        tag1Copy = buildTag("tag1");
+        tag2 = buildTag("tag2");
+    }
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: As we have only one test method using these, may be we should put them inside that method itself?

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 3 above)
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: or just add a `throws Exception` to the method signature. If the exception happens, it will result in a test failure, which is what we want.

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Arrays;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.tag.Tag;
+
+public class UtilsTest {
+    private static Object nullObject = null;
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: declare as constant `NULL_OBJECT`?

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        assertAreUnique(nullObject);
+        assertAreUnique(tag1);
+
+        //all objects unique
+        assertAreUnique("abc", "ab", "a");
+        assertAreUnique(tag1, tag2);
+
+        //some identical objects
+        assertNotUnique("abc", "", "abc");
+        assertNotUnique(tag1, tag1Copy, tag2);
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: In both cases one of the copies is the first object in the list. Chances of finding unexpected bugs can increase if you vary such conditions between test cases. e.g. What if there is a bug in the method that checks uniqueness against the first object only?

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 5 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: ok

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 4 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Ya true should put inside the method which only uses them

**Comment by zzzzwen** [comment]
some more changes

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        // non empty list with one null as the last element
+        assertTrue(Utils.isAnyNull("", new Object(), NULL_OBJECT));
+        assertTrue(Utils.isAnyNull(nonEmptyList, emptyList, NULL_OBJECT));
+    }
+
+    @Test
+    public void elementsAreUnique() throws Exception {
+        Tag tag1 = new Tag("tag1"),
+            tag1Copy = new Tag("tag1"),
+            tag2 = new Tag("tag2");
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Indentation problem

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+        // some identical objects
+        assertNotUnique("abc", "", "abc", "ABC");
+        assertNotUnique(NULL_OBJECT, tag1, tag1Copy, tag2);
+        assertNotUnique("", "abc", "a", "abc");
+        assertNotUnique(NULL_OBJECT, "a", "b", NULL_OBJECT);
+    }
+
+    private void assertAreUnique(Object... objects) {
+        ArrayList<Object> list = new ArrayList<Object>(Arrays.asList(objects));
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Duplicated in line 73. Extract to `createList(Object... objects)`? It can be used in lines 18 and 19 too.

**Comment by damithc** [comment]
@pyokagan can have a look to see if I miss anything?

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import seedu.addressbook.data.tag.Tag;
+
+public class UtilsTest {
+    private static final Object NULL_OBJECT = null;
+
+    @Test
+    public void isAnyNull() {
+        ArrayList<String> emptyList = new ArrayList<String>();
+        ArrayList<String> nonEmptyList = new ArrayList<String>();
+        nonEmptyList.add("A string");
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Maybe we may be testing if `isAnyNull()` peeks inside containers to check if any of their containing elements are `null`. In that case, we should be adding `nulls` into these ArrayLists.

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 8 above)
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Why not just use `Arrays.asList()` directly? They are already `java.util.List`.

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 7 above)
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Should a `commons` test be depending on classes in the ~`model`~ `data` component?

Maybe we should stick with using more basic objects like strings.

If we want to test if `elementsAreUnique()` is calling the `#equals()` and `#hashcode()` methods, then we should use custom stubs for that.

Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: What's the indentation problem?

Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Oh, I didn't notice it's intentional as this form of variable initialization is not common in java. I've seen people recommending one declaration per line e.g. Google's style guide https://google.github.io/styleguide/javaguide.html

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import org.junit.Test;
+
+import seedu.addressbook.data.tag.Tag;
+
+public class UtilsTest {
+    private static final Object NULL_OBJECT = null;
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Yes, using `null` should be fine. If the meaning is clear, it's no longer a 'magic object'.

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 9 above)
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: RE why test empty list, if the SUT is likely to have loops, it's good to throw in cases that can push the loop to its boundaries, such as 0 iterations. It's also confirming that given 0 elements the method returns false, which some developers can get wrong. Also, if I'm doing TDD, that would be my first test case (starting with the simplest case). Of course given the simple implementation of the method some of these cases seem redundant, but I prefer test cases to be written in black box fashion to future proof the tests.

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 7 above)
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Yes, it's best if this class doesn't depend on Model.

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 8 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: This is because the original `elementsAreUnique()` method takes in `Container<E>`, the `java.util.List` class does not implement such interface and thus cannot be used, so I used arraylist

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 9 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: is it necessary to check whether this method goes into the container to check? or just check whether container is not null

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 7 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: I'll remove the test for tags

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 8 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: ya true, I did  that because arraylist more intuitive for me, will change

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 9 above)
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: @damithc Oh, I'm not disputing that :-). I'm just wondering why we are testing with `emptyList` and `nonEmptyList` **ArrayLists**, when `isAnyNull()` takes an `Object...`

Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: > Why do we need to test "empty list", "non empty list"?

Yeah, this is badly worded. Sorry for the confusion.

Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: @zzzzwen If you want to do the "isAnyNull() looks inside containers" check, then this line needs to be
```java
nonEmptyList.add(null);
```

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    public void isAnyNull() {
+        ArrayList<String> emptyList = new ArrayList<String>();
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: If we are doing the "isAnyNull() looks inside containers" thing, then this variable (`emptyList`) is not needed, since it does not tell us anything:

* If `isAnyNull()` looks inside containers, then `isAnyNull(emptyList)` will return false.
* If `isAnyNull()` does not look inside containers, then `isAnyNull(emptyList)` will still return false.

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        ArrayList<String> nonEmptyList = new ArrayList<String>();
+        nonEmptyList.add("A string");
+
+        // empty list
+        assertFalse(Utils.isAnyNull());
+
+        // Any non-empty list
+        assertFalse(Utils.isAnyNull(new Object(), new Object()));
+        assertFalse(Utils.isAnyNull(emptyList));
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: If you want to do the "isAnyNull() looks inside containers" check, then we should use the `nonEmptyList` here.

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+        // non empty list with just one at the beginning
+        assertTrue(Utils.isAnyNull(NULL_OBJECT));
+        assertTrue(Utils.isAnyNull(NULL_OBJECT, "", new Object()));
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: If you want to do tests with different kinds of objects, e.g. strings, containers etc., then you MUST ensure that these objects do not trigger `isAnyNull()`, otherwise we may get false negatives.

e.g. If you want to use `""` here, then in your `// Any non-empty list` tests above, you must add:
```java
assertFalse(Utils.isAnyNull(""));
```

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 10 above)
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Oh, it seems like this is another icky Java thing :-)

Yeah, if we want to call `isAnyNull(null)`, we need to make it explicit by using `isAnyNull((Object)null)`.

**Comment by zzzzwen** [comment]
Some more changes

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 12 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: ok

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    private void assertAreUnique(Object... objects) {
+        assertTrue(Utils.elementsAreUnique(createList(objects)));
+    }
+
+    private void assertNotUnique(Object... objects) {
+        assertFalse(Utils.elementsAreUnique(createList(objects)));
+    }
+
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Trailing whitespace.

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        assertTrue(Utils.isAnyNull(null, new Object(), new Object()));
+
+        // non empty list with nulls in the middle
+        assertTrue(Utils.isAnyNull(new Object(), null, null, "test"));
+        assertTrue(Utils.isAnyNull("", null, new Object()));
+
+        // non empty list with one null as the last element
+        assertTrue(Utils.isAnyNull("", new Object(), null));
+        assertTrue(Utils.isAnyNull(new Object(), new Object(), null));
+
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Trailing whitespace.

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        assertFalse(Utils.isAnyNull());
+
+        // Any non-empty list
+        assertFalse(Utils.isAnyNull(new Object(), new Object()));
+        assertFalse(Utils.isAnyNull("test"));
+        assertFalse(Utils.isAnyNull(""));
+
+        // non empty list with just one at the beginning
+        assertTrue(Utils.isAnyNull((Object)null));
+        assertTrue(Utils.isAnyNull(null, "", new Object(), null));
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: This group says `non empty list with just one at the beginning`, but there is a `null` at the end?

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    private void assertAreUnique(Object... objects) {
+        assertTrue(Utils.elementsAreUnique(createList(objects)));
+    }
+
+    private void assertNotUnique(Object... objects) {
+        assertFalse(Utils.elementsAreUnique(createList(objects)));
+    }
+    
+    private List<Object> createList(Object... objects) {
+        return Arrays.asList(objects);
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: At this point might as well use `Arrays.asList()` directly.

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+        // all objects unique
+        assertAreUnique("abc", "ab", "a");
+        assertAreUnique(1, 2);
+
+        // some identical objects
+        assertNotUnique("abc", "", "abc", "ABC");
+        assertNotUnique(null, 1, new Integer(1));
+        assertNotUnique("", "abc", "a", "abc");
+        assertNotUnique(null, "a", "b", null);
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: To make the tests clearer, maybe you could e.g., for this test `assertNotUnique(null, "a", "b", null)`, add a test above with `assertNotUnique(null, null)`, to make it clear that you are testing that `elementsAreUnique()` still works even if there are unique elements mixed in.

Also, it aids as a check against false negatives (e.g., to check that `elementsAreUnique(null, "a", "b", null)` return false because `elementsAreUnique(null, null)` return false and not because `elementsAreUnique(null, "a")` returns false).

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 16 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: with one at beginning also includes null in the end? or should I split them?

Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: oh I think i should edit this and keep only one at the front

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 17 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: May feel a bit hard to read, I will change it first to see effect

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 18 above)
Comment (zzzzwen) [inline test/java/seedu/addressbook/common/UtilsTest.java]: ok

**Comment by zzzzwen** [comment]
Updated, but I still think that `Arrays.asList()` needs to be separated out

File: test/java/seedu/addressbook/common/UtilsTest.java (hunk 16 above)
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: > oh I think i should edit this and keep only one at the front

Yes

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    private void assertAreUnique(Object... objects) {
+        assertTrue(Utils.elementsAreUnique(Arrays.asList(objects)));
+    }
+
+    private void assertNotUnique(Object... objects) {
+        assertFalse(Utils.elementsAreUnique(Arrays.asList(objects)));
+    }
+}
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Need a newline at end of file.

**Comment by zzzzwen** [comment]
Updated and this should be fine

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+        // empty list
+        assertFalse(Utils.isAnyNull());
+
+        // Any non-empty list
+        assertFalse(Utils.isAnyNull(new Object(), new Object()));
+        assertFalse(Utils.isAnyNull("test"));
+        assertFalse(Utils.isAnyNull(""));
+
+        // non empty list with just one at the beginning
+        assertTrue(Utils.isAnyNull((Object)null));
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Typecast needs to be followed by whitespace e.g. `(Object) null`

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+        assertFalse(Utils.isAnyNull(nullList));
+    }
+
+    @Test
+    public void elementsAreUnique() throws Exception {
+        // empty list
+        assertAreUnique();
+
+        // only one object
+        assertAreUnique((Object)null);
```
Comment (pyokagan) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Typecast needs to be followed by whitespace e.g. `(Object) null`

**Comment by pyokagan** [comment]
@zzzzwen If you're free, maybe port these tests to Addressbook-Level4 as well :-)

**Comment by zzzzwen** [comment]
changed, what do you mean by "port"? Is it a git feature? Or I just copy the test file and create another PR in level4

**Comment by damithc** [comment]
@zzzzwen yes, just copy manually and create a new issue+PR

**Comment by zzzzwen** [comment]
Done for level3 and 4
https://github.com/se-edu/addressbook-level3/pull/59
https://github.com/se-edu/addressbook-level4/pull/234

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+    public void isAnyNull() {
+        // empty list
+        assertFalse(Utils.isAnyNull());
+
+        // Any non-empty list
+        assertFalse(Utils.isAnyNull(new Object(), new Object()));
+        assertFalse(Utils.isAnyNull("test"));
+        assertFalse(Utils.isAnyNull(""));
+
+        // non empty list with just one at the beginning
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: `one null`?

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+
+        // non empty list with nulls in the middle
+        assertTrue(Utils.isAnyNull(new Object(), null, null, "test"));
+        assertTrue(Utils.isAnyNull("", null, new Object()));
+
+        // non empty list with one null as the last element
+        assertTrue(Utils.isAnyNull("", new Object(), null));
+        assertTrue(Utils.isAnyNull(new Object(), new Object(), null));
+
+        // test whether method peeks into the elements inside container
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: Better to be precise
`// confirms nulls inside the list are not considered`

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+        // some identical objects
+        assertNotUnique("abc", "abc");
+        assertNotUnique("abc", "", "abc", "ABC");
+        assertNotUnique("", "abc", "a", "abc");
+
+        assertNotUnique(1, new Integer(1));
+        assertNotUnique(null, 1, new Integer(1));
+
+        assertNotUnique(null, null);
+        assertNotUnique(null, "a", "b", null);
```
Comment (damithc) [inline test/java/seedu/addressbook/common/UtilsTest.java]: It seems the rest of the method has one comment for each code block. Either add comments for these two blocks or merge them with the block above?

**Comment by zzzzwen** [comment]
Updated comments
